lib/std/http/Client.zig

```diff
@@ -1,49 +1,105 @@
-/// managed buffer is not provided.
 allocator: Allocator,
 ca_bundle: std.crypto.Certificate.Bundle = .{},
+ca_bundle_mutex: std.Thread.Mutex = .{},
 /// When this is `true`, the next time this client performs an HTTPS request,
 /// it will first rescan the system for root certificates.
 next_https_rescan_certs: bool = true,
 
+/// The pool of connections that can be reused (and currently in use).
 connection_pool: ConnectionPool = .{},
 
+/// The last error that occurred on this client. This is not threadsafe, do not expect it to be completely accurate.
+last_error: ?ExtraError = null,
+
+pub const ExtraError = union(enum) {
+    fn impliedErrorSet(comptime f: anytype) type {
+        const set = @typeInfo(@typeInfo(@TypeOf(f)).Fn.return_type.?).ErrorUnion.error_set;
+        if (@typeName(set)[0] != '@') @compileError(@typeName(f) ++ " doesn't have an implied error set any more.");
+        return set;
+    }
+
+    // There's apparently a dependency loop with using Client.DeflateDecompressor.
+    const FakeTransferError = proto.HeadersParser.ReadError || error{ReadFailed};
+    const FakeTransferReader = std.io.Reader(void, FakeTransferError, fakeRead);
+    fn fakeRead(ctx: void, buf: []u8) FakeTransferError!usize {
+        _ = .{ buf, ctx };
+        return 0;
+    }
+
+    const FakeDeflateDecompressor = std.compress.zlib.ZlibStream(FakeTransferReader);
+    const FakeGzipDecompressor = std.compress.gzip.Decompress(FakeTransferReader);
+    const FakeZstdDecompressor = std.compress.zstd.DecompressStream(FakeTransferReader, .{});
+
+    pub const TcpConnectError = std.net.TcpConnectToHostError;
+    pub const TlsError = std.crypto.tls.Client.InitError(net.Stream);
+    pub const WriteError = BufferedConnection.WriteError;
+    pub const ReadError = BufferedConnection.ReadError || error{HttpChunkInvalid};
+    pub const CaBundleError = impliedErrorSet(std.crypto.Certificate.Bundle.rescan);
+
+    pub const ZlibInitError = error{ BadHeader, InvalidCompression, InvalidWindowSize, Unsupported, EndOfStream, OutOfMemory } || Request.TransferReadError;
+    pub const GzipInitError = error{ BadHeader, InvalidCompression, OutOfMemory, WrongChecksum, EndOfStream, StreamTooLong } || Request.TransferReadError;
+    // pub const DecompressError = Client.DeflateDecompressor.Error || Client.GzipDecompressor.Error || Client.ZstdDecompressor.Error;
+    pub const DecompressError = FakeDeflateDecompressor.Error || FakeGzipDecompressor.Error || FakeZstdDecompressor.Error;
+
+    zlib_init: ZlibInitError, // error.CompressionInitializationFailed
+    gzip_init: GzipInitError, // error.CompressionInitializationFailed
+    connect: TcpConnectError, // error.ConnectionFailed
+    ca_bundle: CaBundleError, // error.CertificateAuthorityBundleFailed
+    tls: TlsError, // error.TlsInitializationFailed
+    write: WriteError, // error.WriteFailed
+    read: ReadError, // error.ReadFailed
+    decompress: DecompressError, // error.ReadFailed
+};
```

Comment:
this is a bunch of meta garbage and should be deleted

Reply:
The FakeDecompressors were required to work around a false positive dependency loop that made testing impossible, the rest of that is the full set of errors that come as part of the client.

Everything in the http client now returns a small set of curated errors and then dumps the actual error in the `last_error` field. Nearly all of the errors that the layers underneath http can return are likely useless to a user just trying to make a http request.

Reply:
Moving the full set of errors into this union turns:
```
DoError: 134 errors
ReadError: 38 errors
ConnectError: 106 errors
RequestError: 110 errors
```
into 
```
DoError: 24 errors     ( RequestError..., ReadError..., ShortHttpStatusLine, BadHttpVersion,
                         HttpHeadersInvalid, HttpHeaderContinuationsUnsupported,
                         HttpTransferEncodingUnsupported, HttpConnectionHeaderUnsupported,
                         InvalidContentLength, CompressionNotSupported, Uri.ParseError...,
                         TooManyHttpRedirects, HttpRedirectMissingLocation,
                         CompressionInitializationFailed )
ReadError: 4 errors    ( ReadFailed, HttpChunkInvalid, HttpHeadersExceededSizeLimit,
                         OutOfMemory )
ConnectError: 3 errors ( ConnectionFailed, TlsInitializationFailed, OutOfMemory )
RequestError: 7 errors ( ...ConnectError, UnsupportedUrlScheme, UriMissingHost, 
                         CertificateAuthorityBundleFailed, WriteFailed )
```
DoError is longer than the rest because it's the function that handles the most.